### ISSUE TITLE
A bug in proximity_conv()?

### ISSUE THREAD

**Issue by Rasoul77** [robot-learning-freiburg/EfficientLPS#16]
I try to understand the implementation of Proximity Convolution Module where I get errors. Could someone explain to me why `new_knn_idx` is created in the function `proximity_conv()`?
```
  _, knn_idx = difference.topk(
               self.knn_nr, # 9
               dim=1,
               largest=False)
  new_knn_idx = torch.cat((knn_idx,
                           knn_idx + 1 * search_dim,  # knn_idx + 1 * 25
                           knn_idx + 2 * search_dim,  # knn_idx + 2 * 25
                           knn_idx + 3 * search_dim,  # knn_idx + 3 * 25
                           knn_idx + 4 * search_dim), # knn_idx + 4 * 25
                           dim=1 
                          )

```
This causes the error `out of bounds for dimension 1` in the next line when calling the function `torch.gather()`,

`unfold_inputs = torch.gather(input=unfold_inputs, dim=1, index=new_knn_idx)`

**Comment by Rasoul77** [comment]
Here is a minimum working example code,

```
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F


def proximity_conv(inputs, search, knn_nr):
    B, H, W = inputs.shape[0], inputs.shape[-2], inputs.shape[-1]
    search_dim = search ** 2
    pad = int((search - 1) / 2)
    proj_range = inputs[:, 0:1, ...].clone()
    unfold_proj_range = F.unfold(proj_range,
                        kernel_size=(search, search),
                        padding=(pad, pad))
    center = int((search_dim - 1) / 2)
    difference = torch.sqrt((unfold_proj_range - unfold_proj_range[:, center:center + 1, ...]) ** 2)
    difference[:, center:center + 1, ...] = -1
    _, knn_idx = difference.topk(
                    knn_nr,
                    dim=1,
                    largest=False)
    new_knn_idx = torch.cat((knn_idx,
                                knn_idx + 1 * search_dim,  # knn_idx + 1 * 25 (why?)
                                knn_idx + 2 * search_dim,  # knn_idx + 2 * 25 (why?)
                                knn_idx + 3 * search_dim,  # knn_idx + 3 * 25 (why?)
                                knn_idx + 4 * search_dim), # knn_idx + 4 * 25 (why?)
                                dim=1 
                            )
    result = torch.gather(input=unfold_proj_range, dim=1, index=new_knn_idx)

    return result


def main():
    x = np.array([
        [1,2,3,4,5],
        [6,7,8,9,10],
        [11,12,13,14,15],
    ], dtype=np.float32)

    x = np.expand_dims(x, axis=0)
    x = np.expand_dims(x, axis=0)
    x = torch.from_numpy(x)

    result = proximity_conv(x, search=3, knn_nr=9)

if __name__ == "__main__":
    main()

```
